go/tests/qemu_test.go

```diff
@@ -0,0 +1,43 @@
+package tests
+
+import (
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("QEMU", func() {
+
+	pids := []int{}
+
+	It("should start a qemu process with kvm", func() {
+		cmd := exec.Command("qemu-system-x86_64", "--enable-kvm", "-nographic")
+		cmd.Stdout = GinkgoWriter
+		cmd.Stderr = GinkgoWriter
+		Expect(cmd.Start()).To(Succeed())
```

Comment:
I'm not at all familiar with Ginkgo and I clearly need to educate myself, but from the high-level point of view what you're doing here makes sense.

However, you're using `exec.Command` directly instead of the `command.Command` abstraction that we're building for Virt Blocks, which makes the test not as interesting as it could be IMHO.

I notice that you set `Stdout` and `Stderr` to custom Ginkgo versions, which is something we don't currently have hooks for in our abstraction. So I would implement those, and once that's done there should no longer be anything stopping you from adopting our `command.Command`.

Do you want to take care of implementing that functionality as a part of your PR? Or I can get to it myself later, as you prefer.

Reply:
> I'm not at all familiar with Ginkgo and I clearly need to educate myself, but from the high-level point of view what you're doing here makes sense.

You can also choose to write your e2e tests via pure go, I personally like behaviour driven test frameworks for e2e tests. Here more details: https://onsi.github.io/ginkgo/

> 
> However, you're using `exec.Command` directly instead of the `command.Command` abstraction that we're building for Virt Blocks, which makes the test not as interesting as it could be IMHO.
> 
> I notice that you set `Stdout` and `Stderr` to custom Ginkgo versions, which is something we don't currently have hooks for in our abstraction. So I would implement those, and once that's done there should no longer be anything stopping you from adopting our `command.Command`.
> 
> Do you want to take care of implementing that functionality as a part of your PR? Or I can get to it myself later, as you prefer.

I am not sold on the usefulness of `command.Command` yet ;)

Reply:
Oh but I also wanted to add that I will change it.

Reply:
I changed my mind. I looked at the wrapper. It looks like it is a wrapper, just to have a wrapper, without an idea what it is supposed to do or abstract.

I would suggest a more orthogonal approach, having domain-specific code for creating the commandline, creating needed file and so on, and then simply use the built in exec module directly with the output. At least until you know what you want to do with this abstraction.

Reply:
Agreed that the wrapper is not offering much functionality, if any, on top of what you can already get from `exec.Cmd`, and as mentioned earlier is actually missing some. Over the next few days I plan to build a more reasonable API on top of it, where you'd be able to create a "running QEMU process" struct from a `vm.Description`, at which point we'll want to test that API rather than the underlying implementation anyway.

The other thing to keep in mind is that, even if for Go programs using `exec.Cmd` or our very own `command.Command` is basically the same thing, C programs will only have access to the latter through the bindings so it's important that we're testing it: we should standardize on a single way for Virt Blocks to run commands instead of asking users to use the respective native language facilities.

Reply:
> Agreed that the wrapper is not offering much functionality, if any, on top of what you can already get from `exec.Cmd`, and as mentioned earlier is actually missing some. Over the next few days I plan to build a more reasonable API on top of it, where you'd be able to create a "running QEMU process" struct from a `vm.Description`, at which point we'll want to test that API rather than the underlying implementation anyway.
> 

As mentioned before I would rather prefer a more orthogonal approach, where we then don't hide the commandline creation behind an interface, which also runs the process, or monitors it.
There can be from my perspective different granularity for different apis and/or binaries.

> The other thing to keep in mind is that, even if for Go programs using `exec.Cmd` or our very own `command.Command` is basically the same thing, C programs will only have access to the latter through the bindings so it's important that we're testing it: we should standardize on a single way for Virt Blocks to run commands instead of asking users to use the respective native language facilities.

The c-bindings are right now solely for libvirt (at least not for  kubevirt). I can imagine that the bindings will be influenced by that. This may require different APIs than we would want for e.g. kubevirt (and maybe also a different granularity). This is perfectly fine, however from my perspective that just means that more tests are needed, not that we need a standard-api, which will be the only one which we test.

Reply:
> > Agreed that the wrapper is not offering much functionality, if any, on top of what you can already get from `exec.Cmd`, and as mentioned earlier is actually missing some. Over the next few days I plan to build a more reasonable API on top of it, where you'd be able to create a "running QEMU process" struct from a `vm.Description`, at which point we'll want to test that API rather than the underlying implementation anyway.
> 
> As mentioned before I would rather prefer a more orthogonal approach, where we then don't hide the commandline creation behind an interface, which also runs the process, or monitors it.
> There can be from my perspective different granularity for different apis and/or binaries.

I see your point about granularity and agree that it would probably fit well in a project that's all about blocks :)

On the other hand, if your end goal is to run a QEMU process, it would make sense that you'd be able to do that directly instead of having to every time go through the tedious extra steps of taking the generated command line and launching the process yourself.

Since striking a perfect balance would be hard if at all possible, perhaps the best approach then is to provide both APIs so that applications that are served well by a more straightforward API don't have to concern themselves with the details, and applications that require more control can have it.

Does that sound sensible?

> > The other thing to keep in mind is that, even if for Go programs using `exec.Cmd` or our very own `command.Command` is basically the same thing, C programs will only have access to the latter through the bindings so it's important that we're testing it: we should standardize on a single way for Virt Blocks to run commands instead of asking users to use the respective native language facilities.
> 
> The c-bindings are right now solely for libvirt (at least not for kubevirt). I can imagine that the bindings will be influenced by that. This may require different APIs than we would want for e.g. kubevirt (and maybe also a different granularity). This is perfectly fine, however from my perspective that just means that more tests are needed, not that we need a standard-api, which will be the only one which we test.

I don't have a problem with implementing some features purely in the C bindings when Go already has a native facility; for example, I'm doing that already with error handling and arrays.

At the same time, I'm hoping that in most cases we're going to be able to create APIs that makes sense both from Go and from C, and we'll only need to have wildly different APIs exposed to the two languages in some very specific cases. Let's see how well that plays out in practice O:-)

In any case, if we agree on what I wrote above, then we're going to need tests both for the case where the application uses Virt Blocks to generate the command line and then takes care of launching the process itself (the one you have in this patch) and for the case where the application leaves everything to Virt Block. The latter API doesn't exist yet, so we can merge this test case and worry about the other API later.

Reply:
Let's just not think too much about an api for anything right now.

Reply:
Meant "everything", not "anytning".